overlord/devicestate/devicestate.go

```diff
@@ -706,9 +707,13 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 	}
 	snapsAccountedFor[new.Gadget()] = true
 
-	// go through all the model snaps, see if there are new required snaps
-	// or a track for existing ones needs to be updated
-	for _, modelSnap := range new.SnapsWithoutEssential() {
+	// go through the model snaps not accounted for yet, see if there are
+	// new required snaps or a track for existing ones needs to be updated
+	for _, modelSnap := range new.Snaps() {
+		if snapsAccountedFor[modelSnap.Name] {
+			continue
+		}
```

Comment:
I don't think we want to switch snapd track in the middle of other things, it needs to be treated as one of the essential snaps I think and processed first as we do during regular refreshes

Reply:
It'd be really cool if we could do the essential snaps in a loop, rather than having them all listed like they are now. Although some of the special cases make that difficult. Specifically, `new.BaseSnap()` can return `nil` on UC16, and only the gadget snap's task set gets passed to `updateNeededSnapsFromTs`.